Public/Src/IDE/VsCode/BuildXL.IDE.VsCode.dsc

```diff
@@ -1,22 +1,46 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-import {Transformer} from "Sdk.Transformers";
+import {Artifact, Transformer} from "Sdk.Transformers";
 import * as ManagedSdk from "Sdk.Managed";
 import * as Deployment from "Sdk.Deployment";
 import * as Transformers from "Sdk.Transformers";
 import * as DetoursServices from "BuildXL.Sandbox.Windows";
 import * as Branding from "BuildXL.Branding";
 import * as VSIntegration from "BuildXL.Ide.VsIntegration";
+import {Node, Npm} from "Sdk.NodeJs";
 
-namespace LanguageService.Server {
+namespace VsCode.Client {
+    //A new namespace with empty qualifier space to ensure the values inside are evaluated only once
+    export declare const qualifier: {};
+
+    const clientSealDir = Transformer.sealDirectory(d`client`, globR(d`client`));
+
+    const clientCopy: OpaqueDirectory = Deployment.copyDirectory(clientSealDir.root, Context.getNewOutputDirectory("client-copy"), clientSealDir);
+
+    @public
+    export const installRootDir: OpaqueDirectory = Npm.npmInstall(clientCopy);
 
+    @@public
+    export const compileOutDir: OpaqueDirectory = Node.tscCompile(clientCopy.root, clientCopy, installRootDir);
+
+    @@public
+    export const deployedNpmPackageLockFile = Deployment.copyFileFromOpaqueDirectory(

```

Comment:
>deployedNpmPackageLockFile [](start = 17, length = 26)

Don't completely get what this is doing, but do we actually need a special mount for it? Looking at the list of mounts we have defined for our selfhost, this one now stands out as a rare snowflake.

Reply:
We need to check this file into our repo. 

Where do you suggest we place this file?  All other component governance manifest files are under the `cg` folder.  In order to be able to write to that folder from a pip, I thought we had to create a mount for it.

Reply:
Maybe we can copy this package-lock file into an output location, and then use the copied version of it? The fact that we have a writeable place, that is not under Out, and where we also have source files doesn't look super clean to me. At least for now we have always kept outputs and sources well separated.

---
In reply to: [330205562](https://github.com/microsoft/BuildXL/pull/966#discussion_r330205562) [](ancestors = 330205562)

Reply:
This writable mount is defined for `cg/npm` under which there are no source files.

There is nothing we need to use this `package-lock.json` file for; all we need to do is to check it into our repo.  So the question is where it should live.  

Even though the `Out` folder is git ignored, this lock file could live somewhere under `Out` and we could force add it to git.  That, however, seems like a worse solution to me: I don't think any of us would intuitively expect to find any files under `Out` checked into our repo. 

Having a small writable folder with no source files in it (like `cg/npm`) under a source folder (like `cg`) does not seem too bad to me.

Reply:
My proposal was to check in the file where it is in this PR (under cg/npm), but copy the file during the build under some out location, and use the copy of it. In that way you won't need to define any extra mount, and all output files will still be under Out/

---
In reply to: [330240761](https://github.com/microsoft/BuildXL/pull/966#discussion_r330240761) [](ancestors = 330240761)